### DISCUSSION TITLE
Profile Plot Controls?

### DISCUSSION THREAD

**Discussion by BranTruer** [pyswmm/pyswmm#531] [Q&A]
Hello everyone, 
TLDR: I need to know how to control which time series is used for data plotted in a profile view of a SWMM model.  

First time posting here, and I'm new to EPASWMM, Python, pyswmm, swmmio, matplotlib, etc.  I've done programming over the years, first learning FORTRAN many moons ago, but never in a production environment.  

The issue I am having is how to control the data time series, like HGL plotted on the profile.  Going through the tutorials, studying the espresso_machine.py file for clues, I notice there is a blue line plotted in the profile, which I assume is the HGL. but where and what does that line represent.  I cannot find anything anywhere about how to control the time series source of that input.  To be truly meaningful, the values have to all come from the exact same time series, so there must be a way to control the time series being pulled in, but I cannot find it. And I've spent several hours and been to many Web sites looking for it.  It must be right in front of me, but I swear that I do not see any options...so I am posting here asking for help please.  Where do I need to go to find this information? How are you doing this?

I am importing several Python modules: datetime, matplotlib.pyplot, matplotlib.dates, pyswmm, swmmio, pandas, numpy, tabulate, swmm.toolkit, and others. To be sure, my knowledge of these is very limited to the specific things I am working on.

I started the process in EPASWMM, got the model running, then downloaded pyswmm and updated my python-ware.  I got the graphs and reports working, both in EPASWMM and pyswmm.  I also am injecting flows using the pyswmm tools into SWMM, and created factors for each hour, each day, and each month, but also each day of the month.  So far, I feel good about most of this, but I am still debugging the SWMM model, so it would be really nice to have profiles to look at.

Sorry to be so wordy, but I just wanted to assure you that I am not going to waste your time unless I absolutely have done my homework.

Please allow me to thank you ahead of time.

Sincererly

**Comment by bemcdonnell** [comment]
@BranTruer, in that example the profile plotter is rendering the “max” values that show up in the RPT file. If you want to render the timeseries data in a snapshot of time, I think you can pass over a dictionary of depths. You’d have to use pyswmm.Output to pull data from one time step (instead of swmmio to get the max reported depths.)

**Reply by BranTruer** [reply]
@bemcdonnell, thank you for the reply and confirming that is max HGL information!  I kinda figured that out last night, 1st from the name of the swmmio variable being used, and 2nd because it pulls from the report, which only has summary values in it (for some reason that I still do not understand but that's a different topic). 

So I started digging into using pyswmm to get the time series. I've been using ChatGPT to give me some hints on what to look for in the documentation. Now, after many hours,  I've got the code built and I'm at the point of testing and debugging it. Hopefully, this works. I can post some code if interested.

**Reply by bemcdonnell** [reply]
@BranTruer, sorry you feel the docs are hard to follow - when you sort out the problems you’re trying to solve, we would appreciate some detailed feedback so we can improve them.  There is also a YouTube series available on how to use pyswmm. 

The simplest form of timeseries getters using pyswmm.Output is here

http://pyswmm.github.io/pyswmm/reference/api/pyswmm.output.NodeSeries.html 

Hope it helps.

**Reply by BranTruer** [reply]
Ok, thank you, that is helpful!  No apologies needed on the documentation, it is what it is, and I reworded my reply above to be less pointed. 

I did watch and follow the Example videos, thank you for those.  I will incorporate your suggestions. And be happy to give feedback.  thanks again

**Reply by BranTruer** [reply]
@bemcdonnel, I spent the day trying to get this idea about creating a time stamped HGL, but still have not got it.  The main issue, IMHO, is that I do not understand the data structure in Nodes nor in NodesSeries to parse them.  Nor do I understand the interface among pyswmm, swmmio, and matlibplot.  (That is not a passive aggressive attempt to ask for your help.) Even so, it seems to me that the issue boils down to replacing the profile_depths line of code, which currently reads, 
profile_depths = rpt.node_depth_summary.MaxNodeDepthReported.  
And replace that code with code that will set profile_depths equal to time_specific HGL values.

I had dozens and dozens of attempts at this today. 
Those attempts were based mainly on these three ideas, that I feel are close:

Attempt 1. Use keys to parse through the data.
    with Output(r"project.out") as out :
        #for each_node in NodeSeries(out) :
        time_stamp_dict = NodeSeries(out)
        the_keys = time_stamp_dict.keys()
        keys_list = list(the_keys)
        print(keys_list,"," )
        os.system(pause)
        #at this point I could not discover the keys, but the idea would be to use the keys to iterate through NodeSeries
        for the_time_key in keys_list
             hgl_values_at_minute = next(iter(NodeSeries(out)[the_key].hydraulic_head)))
      profile_depths = hgl_values_at_minute

Attempt 2. Parse out the datetime stamps and HGL values for each node.
    with Output(r"project.out") as out :
        for datetimes, hgl_value in NodeSeries(out) :
            if datetimes == DateTime_to_Plot_as_datetime :
                time_stamps.append(datetimes)
                hgl_values_at_minute.append(hgl_value)
    profile_depths = hgl_values_at_minute

Attempt 3. Similar to Attempt 2, but trying to assign the values to a dictionary first, maybe eventually combining this approach with Attempt 1.
    path_selection = find_network_trace(mymodel, startNode, endNode)
    with Output(r"project.out") as out :
        time_series_dict = NodeSeries(out)[int]
        for datetimes, hgl_values in time_series_dict :
            if datetime == DateTime_to_Plot_as_datetime :
                time_stamps.append(datetimes)
                hgl_values_at_minute.append(hgl_value)
    profile_depths = hgl_values_at_minute
#########################################################

So, I am running out of time on this, not sure when I will be able to return to it.  I work for a small consulting engineering firm and we cannot afford to spend much more time on this.  To help with that, I've used 4 weekend days to do this without charging it back to the company. In addition, we set up a meeting for Tuesday, and I still need to go through the SWMM Project's output to verify that I understand it and that we can communicate it to the client. [We are modeling the headworks (influent PS and WW, flow divider, FEB with over flow and drain return) of a WWTP. I have to figure out why I have flow leaking through the bottom of a sharp-crested weir that is on a wall 24" above the floor. And there are a couple of capacity issues, and some preliminary thoughts on where to take the design next.)

The main feedback I have is about the documentation (this is not a complaint): Ignoring that the documentation is partially complete, especially the swimmio functionality.  So for me the struggle is that the documentation is written to much more experienced programmers, hence why ChatGPT was such a big help for me.  Plus the fact that there are so many different data types and object types, that I am challenged to just understand what all of that means.  In other words, I am in over my head, and the documentation barely keeps my head above water.  Again, not a complaint, just feedback.

Thanks again and good luck on this project, it is very impressive and very helpful, especially given the lack in inputting and reporting features that are native to EPASWMM. I get that, they have those engineering models working very well, and that is more than enough of a challenge without working through all of the inputting, outputting, and reporting issues.

If needed, I am happy to share the entire code with you, but this is the essence of it.  
Very little of it is newly written by me, most of it is copy and paste from the tutorials, and ChatGPT, and I served as the editor and secondary progenitor for our project.

Anyway, thanks again.
BT
(obviously a pseudonym, ciao)

**Reply by bemcdonnell** [reply]
@BranTruer can you email me directly?

**Reply by BranTruer** [reply]
yes i'll send you an email

**Comment by BranTruer** [comment]
thank you @bemcdonnell for this answer.  
Here is the code bemcdonnell sent me from PySWMM tutorials.
The key code is about 2/3 way down the page, and indented: "node_depth_dict", 
and then adding it to "add_hgl_plot" a few lines below that.

I was also missing "SystemStats" in the "from pyswmm import"

```ruby
'''
PySWMM Espresso Machine
Author: Bryant McDonnell
Version: 1
Date: May 20, 2024
'''
import warnings
warnings.filterwarnings("ignore")

from datetime import datetime
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from pyswmm import Simulation, Nodes, Links, SystemStats, Output    #<<<missing SystemStats here
from swmm.toolkit.shared_enum import NodeAttribute
import swmmio
from swmmio import find_network_trace
from swmmio import (build_profile_plot, add_hgl_plot, add_node_labels_plot,
                    add_link_labels_plot)


print("Running the Model without Control")
with Simulation(r'DemoModel.inp') as sim:
    sim.step_advance(300)
    for ind, step in enumerate(sim):
        pass

with Output('DemoModel.out') as out:
    node_depth_dict = out.node_attribute(NodeAttribute.INVERT_DEPTH, datetime(2016, 10, 1, 18, 45, 00))  #<<<missing code

mymodel = swmmio.Model(r"DemoModel.inp")
fig = plt.figure(figsize=(11,9))
fig.suptitle("Max HGL")
ax = fig.add_subplot(1,1,1)
path_selection = find_network_trace(mymodel, 'J1', 'J8')
profile_config = build_profile_plot(ax, mymodel, path_selection)
add_hgl_plot(ax, profile_config, depth=node_depth_dict, label="No Control")  #<<<add node_depth_dict here

add_node_labels_plot(ax, mymodel, profile_config)
add_link_labels_plot(ax, mymodel, profile_config)
leg = ax.legend()
ax.grid('xy')
ax.get_xaxis().set_ticklabels([])
```
thank you
BT
